### ISSUE TITLE
Sparsity check upon matrix conversion

### ISSUE THREAD

**Issue by ldrufuca** [bioturing/pyBCS#23]
Trying to convert a valid .h5ad dataset yields the following error
```
>>> scanpy2bcs.format_data("./scanpy_analysis.h5ad", "./data.bcs",
...                         input_format="h5ad", graph_based="leiden_0.6")
Writing main/metadata/metalist.json
Cannot convert tissue to numeric, treating as categorical
Cannot convert disease to numeric, treating as categorical
Cannot convert phase to numeric, treating as categorical
Cannot convert leiden_0.6 to numeric, treating as categorical
"blacklist" is not numeric or categorical, ignoring
Found graph based clustering in metadata with keyword leiden_0.6
Writing main/metadata/bc0d18990a9c4e0f98e44f4a97188b77.json
Writing main/metadata/ddc0df53de60474baf2c4dfd9c29e5ae.json
Writing main/metadata/2e98c4853b4d44b186b79178023cb376.json
Writing main/metadata/c949403b19304eb2beed7c578dfaf54d.json
Writing main/metadata/1353d52a97ef41efa8a37b1ba208f5b1.json
Writing main/metadata/098c15708c994ca0883ec4e117a7fbda.json
Writing main/metadata/df2d0a2ae91e444d85cdc3e3ee177856.json
Writing main/metadata/64776b0d38a34574acc575aea7bdd2ea.json
Writing main/metadata/78239c775e9543ebad32ca3b3dc803c1.json
Writing main/metadata/9b8a52e894a6435797cf919e3aaeacc6.json
Writing main/metadata/c15ccd8c44b94df28ceb9536552130de.json
Writing main/metadata/89d141a71e414519b971dfc330362bce.json
Writing main/metadata/b9618d95f6f74feda83ade574d1c8fa0.json
Writing main/metadata/ff4469ea93d347e9bf1cb75c01e4b6d8.json
Writing main/metadata/8ea1c40cc8004a7993227fe3ba9bb0f0.json
Writing main/metadata/23d67800f0ba4288a55e53382578b4c0.json
Writing main/metadata/93b6d0a44f8942c18e6c8efa66dbdeae.json
Writing main/metadata/e13adc6ce89c4ef2bd35fbf13e77aa6e.json
Writing main/metadata/c0c3567ac3354deebac6c3038fc8a9a6.json
Writing main/metadata/51c1db97e8a344fda55c96610ba5457f.json
Writing main/metadata/c30f8dd93bc44f399d47f918d1aeaccc.json
Writing main/metadata/9ced310fb41c4f6f89d59f4a283baf63.json
Writing main/metadata/5f295632fe8b4c298c65f0a9daeb8f88.json
Writing main/metadata/92e1a93c12094b7698f8a848c2a0821e.json
Writing main/metadata/a8c426ff15a74ef38778a454c04da305.json
Writing main/metadata/903d52a892464fe7a14ae9df54874b43.json
Writing main/metadata/7eb53c0923024d2099e22feaedbb24ce.json
Writing main/metadata/graph_based.json
Writing dimred
Writing X_pca
X_pca has more than 3 dimensions, using only the first 3 of them
Writing X_umap
Writing main/dimred/meta
Writing main/matrix.hdf5
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/storage/home/ldrufuca/miniconda3univa/envs/scanpy160/lib/python3.7/site-packages/pyBCS/scanpy2bcs.py", line 1360, in format_data
    replace_missing=replace_missing)
  File "/storage/home/ldrufuca/miniconda3univa/envs/scanpy160/lib/python3.7/site-packages/pyBCS/scanpy2bcs.py", line 615, in write_bcs
    raise e
  File "/storage/home/ldrufuca/miniconda3univa/envs/scanpy160/lib/python3.7/site-packages/pyBCS/scanpy2bcs.py", line 611, in write_bcs
    self.write_bcs_to_file(zobj, study_name, replace_missing)
  File "/storage/home/ldrufuca/miniconda3univa/envs/scanpy160/lib/python3.7/site-packages/pyBCS/scanpy2bcs.py", line 592, in write_bcs_to_file
    main_path=self.get_main_path(study_name))
  File "/storage/home/ldrufuca/miniconda3univa/envs/scanpy160/lib/python3.7/site-packages/pyBCS/scanpy2bcs.py", line 533, in write_main_folder
    barcodes, features, has_raw = self.write_matrix(dest_hdf5)
  File "/storage/home/ldrufuca/miniconda3univa/envs/scanpy160/lib/python3.7/site-packages/pyBCS/scanpy2bcs.py", line 485, in write_matrix
    norm_M, raw_M, barcodes, features, has_raw = self.get_synced_data()
  File "/storage/home/ldrufuca/miniconda3univa/envs/scanpy160/lib/python3.7/site-packages/pyBCS/scanpy2bcs.py", line 211, in get_synced_data
    norm_M, norm_barcodes, norm_features = self.get_normalized_data()
  File "/storage/home/ldrufuca/miniconda3univa/envs/scanpy160/lib/python3.7/site-packages/pyBCS/scanpy2bcs.py", line 136, in get_normalized_data
    M = self.get_normalized_matrix()
  File "/storage/home/ldrufuca/miniconda3univa/envs/scanpy160/lib/python3.7/site-packages/pyBCS/scanpy2bcs.py", line 673, in get_normalized_matrix
    return self.object.X[:][:].tocsc()
AttributeError: 'numpy.ndarray' object has no attribute 'tocsc'
```
I believe it tracks back to:
https://github.com/bioturing/pyBCS/blob/a9bf0ea01e8865fa423da91ee81861f7430175dd/pyBCS/scanpy2bcs.py#L673

AnnData.X in a fully analyzed dataset is not guaranteed to still be in a sparse format...checking for sparsity may increase robustness? As an alternative, passing sc.read_h5ad(fname, as_sparse='X') might do the trick

**Comment by JATC1024** [comment]
We just released a new version and it should fix this issue. I tested a few examples and it worked. Please let me know if it also works for your case.

**Comment by ldrufuca** [comment]
I tried the new version and it seems to work now
```
>>> scanpy2bcs.format_data("./scanpy_analysis.h5ad", "./data.bcs",
...                         input_format="h5ad", graph_based="leiden_0.4")

Writing main/metadata/metalist.json
Cannot convert tissue to numeric, treating as categorical
Cannot convert disease to numeric, treating as categorical
Cannot convert phase to numeric, treating as categorical
Cannot convert leiden_0.4 to numeric, treating as categorical
"blacklist" is not numeric or categorical, ignoring
Found graph based clustering in metadata with keyword leiden_0.4
Writing main/metadata/4750afea938b4ce1b2debe81b881066b.json
Writing main/metadata/284161179ae34053bd18636c0c01948b.json
Writing main/metadata/89653c4e5d864f74bf48ede6a0165bc5.json
Writing main/metadata/62282df021bc4120a2fd0e89a5cf8829.json
Writing main/metadata/3d9569bace6c4f1c9bf5f4bbdadec877.json
Writing main/metadata/694f96d8a3f646479f4397eec679e563.json
Writing main/metadata/4881fc6715564792b1200e499d6c841e.json
Writing main/metadata/acac4279fae141269bf8ef47c6164fc5.json
Writing main/metadata/f1531434541643aaa27ac29277be74de.json
Writing main/metadata/6202d213e4ab481f93986abf23978072.json
Writing main/metadata/5bebf3e25306466f96ad910d8ff9f650.json
Writing main/metadata/872111e2574040eb82f5a9c23fcbcdf6.json
Writing main/metadata/d2133c8997744121893a1c0c805b612a.json
Writing main/metadata/8a393e39a18b4171b8ade71732412939.json
Writing main/metadata/90f4339b8663435b9df6c30a017fbcb9.json
Writing main/metadata/4b8ab71aa0794e4981be4848a2da3bd1.json
Writing main/metadata/29ab74870a7a4a85b8488ace6b34a796.json
Writing main/metadata/116588e9212f4cd991548313e5c54c89.json
Writing main/metadata/82a75c9eccf04e18902b5d7df65ef4fb.json
Writing main/metadata/33af5df21d0a4da5a8f87650021d1bde.json
Writing main/metadata/e366a5a8ea284b2ea1972fae5bbe6db5.json
Writing main/metadata/8d36129feac644cc81a4d4c8d9aaac4c.json
Writing main/metadata/fe3a7b4c38284a70b03e807fa0e5ef08.json
Writing main/metadata/ad8e1c3e63414187b5040708a97809bb.json
Writing main/metadata/1962314b27bd405e9d01bb700e8b050e.json
Writing main/metadata/0bc9681a3bd94d4d925fcaeb22db4dc7.json
Writing main/metadata/45c2985ac97848fc905263931f909fb1.json
Writing main/metadata/graph_based.json
Writing dimred
Writing X_pca
X_pca has more than 3 dimensions, using only the first 3 of them
Writing X_umap
Writing main/dimred/meta
Writing main/matrix.hdf5
Normalized data and raw data are both available
Writing group "countsT"
Writing group "normalizedT"
Writing group "colsum"
Writing to zip
Writing main/barcodes.tsv
Writing main/genes.tsv
Writing main/gene_gallery.json
Writing run_info.json
```

Nonetheless when I try to load the resulting .bcs file into bbrowser it still fails yielding an error that seems to be related to the formatting of the main matrix

```
[2021-04-16 12:14:11] Extracting data...
[2021-04-16 12:14:14] BCS version: 0.4.1
[2021-04-16 12:14:14] Expanding data...

Error in LightTranspose(file.path(bcs.dir, "main", "matrix.hdf5")) : 
Some error happened, Please post an issue here to report it: https://github.com/bioturing/NoraSC/issues

ERROR
```

**Comment by JATC1024** [comment]
Did you use local mode or server mode when you submitted it to bbrowser?

**Comment by ldrufuca** [comment]
I have the bcs file locally and tried to load it on bbrowser both in local or server mode with no differences in output

**Comment by JATC1024** [comment]
What is the size of your data and how large is the bcs file?

**Comment by ldrufuca** [comment]
dataset is 6012 barcodes, 17321 genes
I am retaining roughly 70 metadatas
The resulting bcs file is 930,7 MB in size

**Comment by JATC1024** [comment]
So it's not big at all. I'll try to figure out the error. It would really help if you send me a chunk of the data (that still produces an error) or a randomized version of it.

**Comment by JATC1024** [comment]
Hi! The matrix stored in the bcs file is actually a minimized version. We do this so that users can pass the bcs files around easier. Therefore, when submitting it to bbrowser, we have to expand it to get the full matrix. The expanding is done in the file bcs_transpose.pyx, which you can find in this repository. According to the log, there was an error when bbrowser tried to expand the matrix, so could you help us do the following:
- Make a copy of the bcs file and change its file extension to .zip
- Unzip it, you will find a file named matrix.hdf5 under the path <unziped folder>/main/matrix.hdf5
- Run python3 bcs_transpose.pyx <path to matrix.hdf5> 100
This will help us with the debugging. Thanks in advance.

**Comment by ldrufuca** [comment]
Hi! I tried running the script, here is the output...there seem to be a mismatch between expected and reconstructed data shape?
```
$ python pyBCS-master/bcs_transpose.pyx c14dfdda-ec31-47c0-9305-ad3a769c24b0/main/matrix.hdf5 100

c14dfdda-ec31-47c0-9305-ad3a769c24b0/main/.2aff3505-affd-4acb-b814-c335b9d8be93
Done a circle in 0.020900 seconds (1.00%)
Done a circle in 0.015130 seconds (1.99%)
Done a circle in 0.018267 seconds (2.99%)
Done a circle in 0.016122 seconds (3.99%)
Done a circle in 0.016708 seconds (4.98%)
Done a circle in 0.016718 seconds (5.98%)
Done a circle in 0.016687 seconds (6.98%)
Done a circle in 0.016287 seconds (7.97%)
Done a circle in 0.018381 seconds (8.97%)
Done a circle in 0.018312 seconds (9.97%)
Done a circle in 0.019195 seconds (10.96%)
Done a circle in 0.017150 seconds (11.96%)
Done a circle in 0.016245 seconds (12.96%)
Done a circle in 0.017303 seconds (13.95%)
Done a circle in 0.017365 seconds (14.95%)
Done a circle in 0.015357 seconds (15.95%)
Done a circle in 0.019873 seconds (16.94%)
Done a circle in 0.016929 seconds (17.94%)
Done a circle in 0.016980 seconds (18.94%)
Done a circle in 0.017744 seconds (19.93%)
Traceback (most recent call last):
  File "pyBCS-master/bcs_transpose.pyx", line 114, in <module>
    bcs_transpose(sys.argv[1], sys.argv[2])
  File "pyBCS-master/bcs_transpose.pyx", line 110, in bcs_transpose
    light_transpose(bcs_path, partial)
  File "pyBCS-master/bcs_transpose.pyx", line 44, in light_transpose
    shape=[n, m])
  File "/Users/ldrufuca/miniconda3/envs/kallisto/lib/python3.7/site-packages/scipy/sparse/compressed.py", line 105, in __init__
    self.check_format(full_check=False)
  File "/Users/ldrufuca/miniconda3/envs/kallisto/lib/python3.7/site-packages/scipy/sparse/compressed.py", line 175, in check_format
    raise ValueError("indices and data should have the same size")
ValueError: indices and data should have the same size
```

**Comment by JATC1024** [comment]
Is it correct that your normalized data has different non-zero entries than your raw data? pyBCS works under the assumption that zero values in the raw matrix stay zero when normalized and vice-versa. Now to think about it, that assumption will be violated all the time. We will work on this and notify you as soon as there is an update.

**Comment by ldrufuca** [comment]
Yes! It really is the case. 
Inspecting my AnnData object I obtain:
```
>>> np.count_nonzero(data.X)
104133852

>>> data.raw.X.count_nonzero()
21887787
```

In my case both matrix have the same shape but that is not guaranteed either (many pipelines hard-filter on highly variable genes in data.X)

**Comment by JATC1024** [comment]
Hi, thanks for waiting. We have published a new version of pyBCS (0.4.2) which handles the case that we discussed. You will also need the newest version of BBrowser (2.9.10) to be able to work with the new version of pyBCS. The update will start automatically the next time you open BBrowser.

pyBCS currently cannot work with datasets where the raw and normalized data have different shapes. This is due to the fact that BBrowser itself cannot work with such data. We may support this in the future, but until then, we will discard the normalized data and keep only the raw data in those cases.
